### PR TITLE
AutoSuspend: Avoid unbalanced prevent/allow Suspend calls

### PR DESCRIPTION
**Pull request by NiLuJe**
This fixes the issue mentioned in https://www.mobileread.com/forums/showthread.php?t=345985, which you can easily reproduce by having autostandby disabled, long-pressing on a book to refresh its metadata, and opening it after the thumbnail has been refreshed.

It would cash a few seconds after, once the metadata processes get reaped.

This also cleans up a few minor things, and one not so minor thing: don't actually suspend if the timeout is unset ;).

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8970)

### CODE REVIEW

**Comment by NiLuJe** [comment]
(Draft because there's still debug prints in, and waiting for @zwim's input to see if there's a slightly less unwieldy option than duplicating the prevent_standby atomic counter).

**Comment by NiLuJe** [comment]
Okay, refactored so that it hopefully makes sense ;).

That only leaves extra debugging to kill, but this is essentially ready for review :).

File: plugins/autosuspend.koplugin/main.lua [hunk 1]
```diff
@@ -142,8 +143,12 @@ function AutoSuspend:onCloseWidget()
     self.task = nil
 
     self:_unschedule_standby()
-    -- allowStandby is necessary, as we do a preventStandby on plugin start
-    UIManager:allowStandby()
+    if not Device:canStandby() then return end
+    -- allowStandby may be necessary, as we do a preventStandby on plugin start
+    while self.prevent_standby_count > 0 do
+        UIManager:allowStandby()
```
Comment (zwim) [inline plugins/autosuspend.koplugin/main.lua]: Shouldn't we use `self:allowStandby()` here? The call here is the counterpart of the initial `self:preventStandby()` called from `_reschedule_standby` through `init`. It seems to me (untested) that using `UIManager:allowStandby` instead of the correct `self:allowStandby` causes the bug.

Then we wouldn't need the `self.prevent_standby_count` here at all.

Comment (zwim) [reply plugins/autosuspend.koplugin/main.lua]: OK, done in [ab90e70](https://github.com/koreader/koreader/pull/8970/commits/ab90e701548d592c1144db0546d2245eb7d3f6c5) :)

File: plugins/autosuspend.koplugin/main.lua [hunk 2]
```diff
@@ -404,7 +429,7 @@ Upon user input, the device needs a certain amount of time to wake up. With some
             help_text = standby_help,
             keep_menu_open = true,
             callback = function(touchmenu_instance)
-                -- 5 sec is the minimum and 60*60 sec (15min) is the maximum standby time.
+                -- 3 sec is the minimum and 15*60 sec (15min) is the maximum standby time.
                 -- We need a minimum time, so that scheduled function have a chance to execute.
                 -- A standby time of 15 min seem excessive.
                 -- But or battery testing it might give some sense.
```
[the remark this answers is not in the document]
Comment (NiLuJe) [reply plugins/autosuspend.koplugin/main.lua:435]: Probably makes sense anyway, although the timeout has already passed by the time onAllowStandy runs (as allowStandby is what lets UIManager send the AllowStandy event).

**Comment by zwim** [comment]
Can not test right now. But it looks good.

File: plugins/autosuspend.koplugin/main.lua [hunk 3]
```diff
@@ -470,30 +470,29 @@ function AutoSuspend:onAllowStandby()
...
         local wake_in = math.huge
         -- The next scheduled function should be the deadline_guard
-        -- Wake before the second next scheduled function executes (e.g. footer update, suspend ...)
+        -- Wake up before the second next scheduled function executes (e.g. footer update, suspend ...)
         local scheduler_times = UIManager:getNextTaskTimes(2)
         if #scheduler_times == 2 then
             -- Wake up slightly after the formerly scheduled event, to avoid resheduling the same function
-            -- after a fraction of a second again (e.g. don't draw footer twice)
+            -- after a fraction of a second again (e.g. don't draw footer twice).
             wake_in = math.floor(scheduler_times[2]:tonumber()) + 1
```
Comment (zwim) [inline plugins/autosuspend.koplugin/main.lua:478]: After uppercase. nit

**Comment by zwim** [comment]
Just a thought to be consistent: Would renaming all xxx_tv which use elapsedTimeSinceBoot to xxx_btv make sense?

File: plugins/autosuspend.koplugin/main.lua [hunk 4]
```diff
@@ -421,17 +446,22 @@ end
...
+    -- see its own AllowStandby handler for more details.
+
+    -- Don't enter standby if we haven't set a proper timeout yet.
+    if not self:_enabledStandby() then
+        logger.dbg("AutoSuspend: No timeout set, no standby")
+        return
+    end
 
-    -- Don't enter standby if wifi is on, as this my break reconnecting (at least on Kobo-Sage)
+    -- Don't enter standby if wifi is on, as this will break in fun and interesting ways (from Wi-Fi issues to kernel deadlocks).
```
Comment (zwim) [inline plugins/autosuspend.koplugin/main.lua:458]: Those three guards (if ...) could go into one if elseif elseif block.

File: plugins/autosuspend.koplugin/main.lua [hunk 5]
```diff
@@ -440,30 +470,29 @@ function AutoSuspend:onAllowStandby()
     if Device:canStandby() then
```
Comment (zwim) [inline plugins/autosuspend.koplugin/main.lua:470]: And a last
```
 elseif not Device:canStandby()
     return
 end
```
could make the code cleaner.

File: plugins/autosuspend.koplugin/main.lua (hunk 4 above)
Comment (NiLuJe) [reply plugins/autosuspend.koplugin/main.lua:458]: They're not mutually exclusive though ;).

**Comment by NiLuJe** [comment]
> Just a thought to be consistent: Would renaming all xxx_tv which use elapsedTimeSinceBoot to xxx_btv make sense?

It's not necessarily cheese enough (or worth the distinction), as I dint think we really have anything that uses timeval w/ real-time.

The casual reader doesn't really care about the distinction between monotonic & boottime ;) '

**Comment by NiLuJe** [comment]
@zwim: I'm planning on merging this one this week-end if you want to give it a try, FWIW (probably Sunday or Monday, as I'm working this Saturday).

**Comment by zwim** [comment]
I will test it on saturday.


Am 8. April 2022 00:37:54 MESZ schrieb NiLuJe ***@***.***>:
***@***.***: I'm planning on merging this one this week-end if you want to give it a try, FWIW (probably Sunday or Monday, as I'm working this Saturday).
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/koreader/koreader/pull/8970#issuecomment-1092273568
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by zwim** [comment]
One minor problem consists (not new). 
When setting autostandby_timout to 4sec and then trigger a rerendering which takes longer than 4sec, then autostandby get's unscheduled until the next tap.

Maybe we should add an appropriate eventHandler (onUpdatePos (?) ) to reschedule standby:
```
function AutoSuspend:onUpdatePos()
    self:_reschedule_standby()
end
```

**Comment by NiLuJe** [comment]
Or we could let it lie, it doesn't feel like a problematic corner case :).

**Comment by zwim** [comment]
It might not be problematic, but my tests show, that a reschedule in onUpdatePos solves the problem.

**Comment by NiLuJe** [comment]
My issue with that is that the actual corner case is very edgy, while the
workaround will effectively be duplicating onInputEvent 99.999% of the rule
;).

**Comment by NiLuJe** [comment]
Saw the edit: I don't recall if there's something closer related to settings updates (post rendering).

**Comment by zwim** [comment]
Just a thought: In onAllowStandby we check if the next standby is scheduled to be longer than 3 seconds, if not we skip and loose the scheduling. This could be the real cause and than this behavior is not a corner case, but will hit us every time when a tap is less than 3 secs before a footer update as well.

Should we add an else branch with a new scheduling?

```
    if wake_in > 3 then -- don't go into standby, if scheduled wakeup is in less than 3 secs
        UIManager:broadcastEvent(Event:new("EnterStandby"))
        logger.dbg("AutoSuspend: going to standby and wake in " .. wake_in .. "s zZzzZzZzzzzZZZzZZZz")

        -- This obviously needs a matching implementation in Device, the canonical one being Kobo.
        Device:standby(wake_in)

        logger.dbg("AutoSuspend: leaving standby after " .. Device.last_standby_tv:tonumber() .. " s")

        UIManager:broadcastEvent(Event:new("LeaveStandby"))
        self:_unschedule() -- unschedule suspend and shutdown, as the realtime clock has ticked
        self:_schedule()   -- reschedule suspend and shutdown with the new time
        -- Don't do a `self:_reschedule_standby()` here, as this will interfere with suspend.
        -- Leave that to `onLeaveStandby`.
    else
        self:_reschedule_standby()
    end
```

**Comment by poire-z** [comment]
Btw, do we really, at each input event, unschedule() and re-scheduleIn() ?
I think in some other possibly similar thingies in the code, we just update some `should_standby_at_tv` - and the previously scheduled action, when executed, would check if that's in the past (in which case: standby) or in the future, in which case it would reschedule itself to be run in `should_standby_at_tv - now()`.
Just mentionning that in case it could make things leaner.

**Comment by zwim** [comment]
@NiLuJe 
(Not closely connected to this PR, but at least a bit due to @poire-z scheduling hint.)
I was looking into some method to speed up UIManagers schedule function (which I must confess is really good).
My first thoughts were about the used data structures. If we could get rid of TimeVal and use numbers (seconds with fractions) here, the scheduler speed would improve massively (with 80 scheduled entries by the factor of 110).

* Why do we use TimeVals here in Lua-lan. Luajit uses doubles as number. A double has a 52bit mantissa plus a sign bit, which allows to save 2^52 different numbers (0 to 4x10e15).
* On the other hand our TimeVal data comes from the OS and stores at maximum the seconds since 1970 (maybe less with clock_monotonic, clock_boottime). A year has 3x10e7 seconds and 100 years have 3x10e9 seconds (that would be at worst 2070 at best 100 years after boot -> in the year 2122 (wouldn't that be a nice song: in the year 2122..2122)).
* So, if we would store the TimeVal in a double we have 15-9 = 6 digits room for the subsecond part. And that is exactly our us precision now.
* If KOReader lives after 2070 and the underlying number-format of Lua is not changed, then we could use a precision of 10us until the year 2970. Do we use times less than 10us anywhere now? Maybe Lua upgrades it numbers to long double and we won't have any problems until the year 12000 (guess).

So my question is: Why don't we want to get rid of the TimeVal data structure in KOReader and use a more efficient data structure based on numbers? (I am aware that 1.1 seconds is not exactly stored as 1.1 due to the IEEE format, but if this is a problem a  fixed-point format might be used, e.g. time = sec*1e6+usec with 1us prec. or time=sec*1e7+usec for +100 years compatibility.)

Pros: 
* `UIManager:schedulexxx`  gets really faster.
* All calculations of time differences get faster.
* All calculations of times in the future get faster.
* Faster means power savings in all situations above.
Cons:
* With our micro-seconds precision we are only fit until 2070 (but 10us we could go until 2970).

**Comment by Frenzie** [comment]
I've had the same thought before, but note that I would (at least as a starting point) phrase that as "completely change how TimeVal works internally". Any similarity in naming to sys/time.h is purely coincidental.

**Comment by NiLuJe** [comment]
That would imply using a posix timeval data type via ffi, if I follow?

**Comment by NiLuJe** [comment]
As for using Lua numbers, in increasing order of concerns are the range, the precision, but, mainly, the small and nasty rounding issues that subtly break a lot of crap ;).

If you have some code no hand, I'd be interested in more details on your fixed-point approach ;).

**Comment by zwim** [comment]
@NiLuJe You can find an implementation in https://github.com/zwim/koreader/tree/changeTimeVal.
This tree is just a proof of concept.

Changes there are using `fts` (fixed-point-time-seconds with µs precision) in the `schedule` and `_checkTask` functions, as well as two methods to convert to and from TimeVal.
The goal would be to get rid of TimeVal at all and use `fts` everywhere in KOReader.

**Comment by zwim** [comment]
For you concerns in descending order:
* There is no rounding. A value of 1 in an `fts` means 1µs no rounding, 10 is 10µs 1 000 000 is 1 s. 
* Precision is 1µs.
* We are safe for 285 years (1970+285 = 2255, almost a precision landing https://www.youtube.com/watch?v=vdSqLfuRN18)

**Comment by NiLuJe** [comment]
@zwim: That looks good enough on paper ;).

Feel free to try it out for real (possible `new` becomes a simple method that takes `sec, usec` args and returns a plain untagged number, and we lose the table and metatable?) ;).

**Comment by NiLuJe** [comment]
One possible hiccup with losing the (posix) timeval-like struct data structure is Input, which relies on that *a lot*. Not entirely sure making scheduling faster at the cost of making input slower because would be a net win ;).

**Comment by NiLuJe** [comment]
In any case, going to merge this as-is, because that how it's been tested, and I'll play with the rescheduling trick in a new branch ;).
